RDSC-4603 Update RDI public documentation (#3262)
redis/docs: content/integrate/redis-data-integration/installation/migration-classic-to-flink.md

and currently runs on Kubernetes only. It can achieve much higher throughput 
during snapshots, scales horizontally by changing the number of TaskManager replicas, 
and uses Flink checkpointing for fault tolerance. See [Stream processor implementations]({{< relref "/integrate/redis-data-integration/architecture#stream-processor-implementations" >}})
for an overview.

This page describes how to migrate an existing pipeline from the classic
processor to the Flink processor.

{{< note >}}The Flink processor is currently supported on Kubernetes only. VM
installations must continue to use the classic processor.{{< /note >}}

## Before you migrate

Confirm that your pipeline is compatible with the Flink processor:

-   The Flink processor supports `hash` and `json` target data types only. If
    any of your jobs use the `set`, `sorted_set`, `stream`, or `string` data
    types, those jobs must be rewritten or kept on the classic processor.
-   `JSON.MERGE` semantics differ from the classic processor's Lua-based merge
    when null values are involved (see
    [`use_native_json_merge`]({{< relref "/integrate/redis-data-integration/reference/config-yaml-reference#processors" >}})).

>     The Flink processor always uses the native `JSON.MERGE` command when the

Reviewer: @stoyanr @andy-stark-redis  I'm not sure if this is clear enough (to me it is, but for users it might not be).

I think we need to explicitly say that:
```
The classic processor JSON.MERGE and Flink processor JSON.MERGE work the same way
```

I.e. the classic processor supports both:
* Lua-based merge
* JSON.MERGE

The flink processor supports only:
* JSON.MERGE

Author: The Flink processor would always use the native `JSON.MERGE` if supported, and would fallback to code-based merge if not (so the merge strategy is supported also in such cases). This is similar to the classic processor, except that `use_native_json_merge` setting is not supported - there is no way to disable native JSON merge if supported by Redis. So I think the docs here are correct; let me know if you think this should be phrased differently.

> -   `JSON.MERGE` semantics differ from the classic processor's Lua-based merge
>     when null values are involved (see
>     [`use_native_json_merge`]({{< relref "/integrate/redis-data-integration/reference/config-yaml-reference#processors" >}})).
>     The Flink processor always uses the native `JSON.MERGE` command when the
>     target database supports it.

Reviewer 2: @stoyanr @ZdravkoDonev-redis Does this summarise the issue OK?
```suggestion
 -   The classic processor merges JSON objects using Lua scripts by default, but you can optionally enable the [`use_native_json_merge`]({{< relref "/integrate/redis-data-integration/reference/config-yaml-reference#processors" >}}) setting to use the Redis [`JSON.MERGE`]({{< relref "/commands/json.merge" >}}) command instead. The Flink processor only supports merging with `JSON.MERGE`. The Lua scripts handle null values in a different way to `JSON.MERGE`, so if you are using the Lua approach with the classic processor, you should check for any problems that this might cause before you migrate.
```

Author: No, it does not, see my previous comment. I think what is said in this section is exactly correct; it could be rephrased, but its meaning should not be changed. 
* `use_native_json_merge` defaults to true also for the classic processor; so without this setting, the 2 processors behave exactly in the same way - they use JSON.MERGE if supported and fallback to a code-based merge if not (the Flink processor does it via Java code instead of Lua but that's an impl detail).
* The only actual difference is that you can set `use_native_json_merge` to`false` for the classic processor, but not for the Flink processor.

Reviewer: Yeah ok. I agree with Stoyan

and starts the Flink JobManager and TaskManager workloads for the pipeline.

## Step 3: Adapt deprecated and Classic-only properties

Some `processors` properties are no-ops, classic-only, or have moved to
`processors.advanced` for the Flink processor. The following table lists the
properties that need attention when migrating.

| Property | Action when migrating to Flink |
| :-- | :-- |
| `on_failed_retry_interval` | No-op. Remove. |
| `duration` | No-op. Use `read_batch_timeout_ms` instead. |
| `dedup`, `dedup_max_size`, `dedup_strategy` | Classic-only. Remove. |
| `enable_async_processing`, `batch_queue_size`, `ack_queue_size` | Classic-only. Remove. |
| `initial_sync_processes` | Classic-only. Configure parallelism through `advanced.flink.taskmanager.numberOfTaskSlots` and `advanced.resources.taskManager.replicas` instead. |
| `idle_streams_check_interval_ms`, `busy_streams_check_interval_ms` | Classic-only. Use `processors.advanced.source.discovery.interval.ms` for a single discovery interval. |
| `idle_sleep_time_ms` | Classic-only. Remove. |
| `use_native_json_merge` | Classic-only. The Flink processor always uses `JSON.MERGE` when the target supports it. |

> The classic processor silently ignores `processors.advanced`, so keeping

Reviewer: @stoyanr @andy-stark-redis Perhaps it's not clear from this paragraph here so we should also explicitly say that the Flink processor ignores the classic-only properties as well?

Author: Added

both top-level properties and their `processors.advanced` equivalents lets

> you switch back without further edits.

Reviewer 2: ```suggestion
you switch back without further edits. Likewise, the Flink processor silently ignores the settings that only apply under classic.
```

Author: I proposed a slightly different sentence, hopefully still ok.

## Step 4: Tune the Flink processor (optional)

Fine-tune the Flink processor through the `processors.advanced` section.
For example:

```yaml
processors:
  type: flink
  advanced:
    source:
      # Time between checks for new input streams.
      discovery.interval.ms: 1000
    target:
      # Verify writes are replicated before acknowledging.
      wait.enabled: true
      wait.write.timeout.ms: 1000
    flink:
      # Number of parallel task slots per TaskManager pod.
      taskmanager.numberOfTaskSlots: 2
      # Total memory budget for each TaskManager JVM process.
      taskmanager.memory.process.size: 4096m
    resources:
      taskManager:
        # Number of TaskManager pods
        replicas: 2
```

See the
[`processors.advanced` reference]({{< relref "/integrate/redis-data-integration/reference/config-yaml-reference#processors" >}})
for the full set of available properties.

## Step 5: Update observability

> The Flink processor does not use `rdi-metrics-exporter`. It exposes

Reviewer: ```suggestion
The Flink processor does not use `rdi-metrics-exporter` component. It exposes
```

Author: I changed the docs to talk about "`rdi-metrics-exporter` service" everywhere consistently. This particular reference here I dropped as it was not needed.
